### PR TITLE
Add example showing how to retrieve a checkpoint from an xgboost model.

### PR DESCRIPTION
**Pull request by geoalgo**
*Issue #, if available:* https://github.com/awslabs/syne-tune/issues/422

*Description of changes:* Provides an example to retrieve the best checkpoint of a model after tuning. Also show log of trials to user when no metrics are emitted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/examples.rst [hunk 1]
```diff
@@ -207,6 +207,19 @@ are strongly recommended for multi-fidelity HPO with the SageMaker backend.
...
 `this tutorial <tutorials/benchmarking/bm_sagemaker.rst#using-sageMaker-managed-warm-pools>`__.
 
+Retrieving the best checkpoint
+=================================
+
+.. literalinclude:: ../../examples/launch_standalone_bayesian_optimization.py
+   :caption: examples/launch_standalone_bayesian_optimization.py
+   :lines: 13-
+
+Syne Tune combines a scheduler (HPO algorithm) with a backend to provide a
```
Comment (mseeger) [inline docs/source/examples.rst:217]: Copy and pasted from another paragraph? Also, does not include your example

File: docs/source/faq.rst [hunk 2]
```diff
@@ -269,6 +269,14 @@ Finally, the scheduler provides additional information about checkpointing in
 this: ``add_checkpointing_to_argparse(parser)`` adds corresponding arguments to
 the parser.
 
+
+How can I retrieve the best checkpoint obtained after tuning?
+===========================================
```
Comment (mseeger) [inline docs/source/faq.rst:274]: Doesn't this === need to be as long as the line above?

File: examples/launch_checkpoint_example.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+"""
```
Comment (mseeger) [inline examples/launch_checkpoint_example.py:1]: Below the copyright comments?

File: examples/launch_checkpoint_example.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        scheduler=BayesianOptimization(config_space, metric="merror", mode="min"),
+        stop_criterion=StoppingCriterion(max_wallclock_time=10),
+        n_workers=n_workers,
+    )
+
+    tuner.run()
+
+    exp = load_experiment(tuner.name)
+    best_config = exp.best_config()
+    checkpoint = exp.path / str(best_config["trial_id"]) / "checkpoints"
```
Comment (mseeger) [inline examples/launch_checkpoint_example.py:58]: Are you sure the config always has this "trial_id" field?

Comment (mseeger) [reply examples/launch_checkpoint_example.py:58]: I also think it would be better to define some helper function, maybe `checkpoint_path(trial_id)`, which is then both called by `LocalBackend` and by this example here. This allows us to change things later on.

File: examples/training_scripts/xgboost/xgboost_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    loaded.load_model(path)
+    return loaded
+
+
+def evaluate_accuracy(checkpoint_dir):
+    X, y = load_digits(return_X_y=True)
+
+    clf = load_model(checkpoint_dir=checkpoint_dir)
+    y_pred = clf.predict(X)
+    return (np.equal(y, y_pred) * 1.0).mean()
```
Comment (mseeger) [inline examples/training_scripts/xgboost/xgboost_checkpoint.py:86]: Why this * 1.0 ?

File: examples/training_scripts/xgboost/xgboost_checkpoint.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+import numpy as np
+import xgboost
+from sklearn.datasets import load_digits
+
+from syne_tune import Reporter
+from syne_tune.constants import ST_CHECKPOINT_DIR
+
+
+class SyneTuneCallback(xgboost.callback.TrainingCallback):
```
Comment (mseeger) [inline examples/training_scripts/xgboost/xgboost_checkpoint.py:25]: Cool to have an xgboost example

File: examples/training_scripts/xgboost/xgboost_checkpoint.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    parser = ArgumentParser()
+    parser.add_argument("--max_depth", type=int, required=False, default=1)
+    parser.add_argument("--gamma", type=float, required=False, default=2)
+    parser.add_argument("--reg_lambda", type=float, required=False, default=0.001)
+    parser.add_argument("--n_estimators", type=int, required=False, default=10)
+    parser.add_argument(f"--{ST_CHECKPOINT_DIR}", type=str, default="./")
+
+    args, _ = parser.parse_known_args()
+
+    checkpoint_dir = Path(vars(args)[ST_CHECKPOINT_DIR])
```
Comment (mseeger) [inline examples/training_scripts/xgboost/xgboost_checkpoint.py:99]: getattr(args, ST_CHECKPOINT_DIR) ?

File: syne_tune/experiments.py [hunk 8]
```diff
@@ -46,6 +46,7 @@ class ExperimentResult:
     results: pd.DataFrame
     metadata: dict
     tuner: Tuner
+    path: Path
```
Comment (mseeger) [inline syne_tune/experiments.py:49]: Please also extend the docstring.

BTW: Is this a fix, or something that is needed for this example?

File: docs/source/examples.rst (hunk 1 above)
Comment (geoalgo) [reply docs/source/examples.rst:217]: That's a type, it is fixed.

File: docs/source/faq.rst (hunk 2 above)
Comment (geoalgo) [reply docs/source/faq.rst:274]: Did not know, fixed it.

File: syne_tune/experiments.py (hunk 8 above)
Comment (geoalgo) [reply syne_tune/experiments.py:49]: It is needed for this example to get the path of experiment.

File: examples/training_scripts/xgboost/xgboost_checkpoint.py (hunk 7 above)
Comment (geoalgo) [reply examples/training_scripts/xgboost/xgboost_checkpoint.py:99]: Is there an advantage of using getattr?

File: examples/training_scripts/xgboost/xgboost_checkpoint.py (hunk 5 above)
Comment (geoalgo) [reply examples/training_scripts/xgboost/xgboost_checkpoint.py:86]: It allows to compute the average with np.mean.

File: examples/launch_checkpoint_example.py (hunk 4 above)
Comment (geoalgo) [reply examples/launch_checkpoint_example.py:58]: > Are you sure the config always has this "trial_id" field?

Yes, this field is always added.

Comment (geoalgo) [reply examples/launch_checkpoint_example.py:58]: > I also think it would be better to define some helper function, maybe checkpoint_path(trial_id), which is then both called by LocalBackend and by this example here. This allows us to change things later on.

Thanks, I updated with this suggestion.

File: examples/training_scripts/xgboost/xgboost_checkpoint.py (hunk 7 above)
Comment (mseeger) [reply examples/training_scripts/xgboost/xgboost_checkpoint.py:99]: Not so important, but vars(args) probably creates a full dict from, which you just use one entry.

**Review by mseeger — APPROVED** [review]
